ingestion-edge/ingestion_edge/disk_queue.py

```diff
@@ -6,65 +6,93 @@
-            with write_lock:
-                # TODO rotate queue file and handle rotated files first
-                with open(QUEUE_FILE, "r") as q:
-                    for line in q:
-                        if end < start:
-                            # TODO use q.seek(start)
-                            end += 1
-                            continue
-                        client.publish(**ujson.loads(line)).result()
-                        # update end only on success
-                        end += 1
-            remove(QUEUE_FILE)
-            done = True
+            # process queue
+            # TODO rotate queue file and handle rotated files first
+            # NOTE rotated files will not require write_lock
+            with open(QUEUE_FILE, "r") as q:
+                # seek to previous end position
+                q.seek(start)
+                # can't use `for line in q` because it disables `q.tell()`
+                # and doesn't allow for releasing write_lock between line reads
+                # or using formats with framing other than newline
+                while True:
+                    # write_lock ensures we don't read a partial line
+                    # or write to queue immediately before removing it
+                    with write_lock:
+                        line = q.readline()
+                        if not line:
+                            # at EOF, remove queue and break
+                            # ignore FileNotFoundError
+                            # remove position first to prevent corrupted state
+                            _remove_if_found(QUEUE_POSITION_FILE)
+                            _remove_if_found(QUEUE_FILE)
+                            done = True
+                            break
+                    # when this throws an exception, stop flushing
+                    # TODO return 202 for transient exceptions
+                    client.publish(**ujson.loads(line)).result()
+                    # record new file position
+                    end = q.tell()
+                    # increment counter for items successfully published
+                    published += 1
+        except FileNotFoundError:
+            # no queue to flush
+            return "", 204
         finally:
-            # update the queue position
-            if end > start and not done:
+            # update queue position if we didn't finish but made progress
+            if not done and end > start:
+                # use upside down floor division for ceiling division
+                end_length = -(-end.bit_length() // 8)
+                end_bytes = end.to_bytes(end_length, sys.byteorder)
                 # record queue position if we moved forward and didn't finish
-                with open(QUEUE_POSITION_FILE, "r+") as qp:
-                    qp.write(str(end))
+                with open(QUEUE_POSITION_FILE, "wb") as qp:
+                    qp.write(end_bytes)
```

Comment:
I think it is an improvement to seek based on byte position instead of line number, but I'm not as convinced that storing the queue position as bytes instead of text is a win. It seems more complicated, and would make it a lot harder to read the queue file by hand if it came to that. It saves a few bytes, but the open/read/close loop is surely dominated by opening the file, not reading a few bytes.